File: operator/diy-project-blogs/projects/clustered-agent-benchmark-uncertainty/render-figure.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const root = path.dirname(fileURLToPath(import.meta.url));
const results = JSON.parse(fs.readFileSync(path.join(root, "results.json"), "utf8"));
const scenarioLabels = new Map([
  ["heterogeneity-0", "No cluster heterogeneity"],
  ["heterogeneity-0.1", "Moderate heterogeneity"],
  ["heterogeneity-0.2", "Strong heterogeneity"],
  ["clusters-10", "10 repositories"],
  ["clusters-20", "20 repositories"],
  ["clusters-80", "80 repositories"],
]);
const methodLabels = {
  naive: "Task normal",
  taskBootstrap: "Task bootstrap",
  clusterBootstrap: "Repository bootstrap",
};
const colors = {
  naive: "#ff8a65",
  taskBootstrap: "#f6c453",
  clusterBootstrap: "#5eead4",
};

function escapeXml(value) {
  return String(value)
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;");
}

function panel(scenarios, x, y, width, height, title) {
  const top = y + 58;
  const bottom = y + height - 54;
  const chartHeight = bottom - top;
  const yMin = 0.88;
  const yMax = 0.97;
  const projectY = (value) => bottom - ((value - yMin) / (yMax - yMin)) * chartHeight;
  const groupWidth = (width - 94) / scenarios.length;
  const marks = [];

  marks.push(`<rect x="${x}" y="${y}" width="${width}" height="${height}" rx="10" fill="#111b32" stroke="#2b3b5c"/>`);
  marks.push(`<text x="${x + 24}" y="${y + 34}" class="panel-title">${escapeXml(title)}</text>`);
  for (const tick of [0.90, 0.92, 0.94, 0.95, 0.96]) {
    const tickY = projectY(tick);
    const dash = tick === 0.95 ? "" : ' stroke-dasharray="4 5"';
    const stroke = tick === 0.95 ? "#dbeafe" : "#344464";
    marks.push(`<line x1="${x + 58}" y1="${tickY}" x2="${x + width - 20}" y2="${tickY}" stroke="${stroke}"${dash}/>`);
    marks.push(`<text x="${x + 48}" y="${tickY + 4}" class="axis" text-anchor="end">${Math.round(tick * 100)}%</text>`);
  }

  scenarios.forEach((scenario, index) => {
    const center = x + 70 + groupWidth * index + groupWidth / 2;
    const entries = Object.entries(scenario.methods);
    entries.forEach(([method, metrics], methodIndex) => {
      const cx = center + (methodIndex - 1) * 24;
      const cy = projectY(metrics.coverage);
      const error = metrics.coverageMonteCarloSE * 1.96;
      const high = projectY(Math.min(yMax, metrics.coverage + error));
      const low = projectY(Math.max(yMin, metrics.coverage - error));
      marks.push(`<line x1="${cx}" y1="${high}" x2="${cx}" y2="${low}" stroke="${colors[method]}" stroke-width="2"/>`);
      marks.push(`<line x1="${cx - 5}" y1="${high}" x2="${cx + 5}" y2="${high}" stroke="${colors[method]}" stroke-width="2"/>`);
      marks.push(`<line x1="${cx - 5}" y1="${low}" x2="${cx + 5}" y2="${low}" stroke="${colors[method]}" stroke-width="2"/>`);
      marks.push(`<circle cx="${cx}" cy="${cy}" r="6" fill="${colors[method]}"/>`);
    });
    const label = scenarioLabels.get(scenario.label) ?? scenario.label;
    const words = label.split(" ");
    const split = words.length > 2 ? Math.ceil(words.length / 2) : words.length;
    marks.push(`<text x="${center}" y="${bottom + 25}" class="label" text-anchor="middle">${escapeXml(words.slice(0, split).join(" "))}</text>`);
    if (split < words.length) marks.push(`<text x="${center}" y="${bottom + 42}" class="label" text-anchor="middle">${escapeXml(words.slice(split).join(" "))}</text>`);
  });
  return marks.join("\n");
}

const heterogeneity = results.scenarios.filter((scenario) => scenario.label.startsWith("heterogeneity"));
const clusters = results.scenarios.filter((scenario) => ["clusters-10", "clusters-20", "heterogeneity-0.2", "clusters-80"].includes(scenario.label));
clusters.sort((a, b) => a.repositoryCount - b.repositoryCount);

const legend = Object.entries(methodLabels).map(([method, label], index) => {
  const x = 166 + index * 260;
  return `<circle cx="${x}" cy="129" r="6" fill="${colors[method]}"/><text x="${x + 14}" y="134" class="legend">${escapeXml(label)}</text>`;
}).join("\n");

const svg = `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1280 720" width="1280" height="720" data-visual-quality="publication" data-text-fit="bounded" role="img" aria-labelledby="title desc">
<title id="title">Confidence interval coverage under repository-clustered coding tasks</title>
<desc id="desc">Two-panel result chart showing 95 percent interval coverage for task-normal, task-bootstrap, and repository-bootstrap estimators as treatment heterogeneity and repository count change. Error bars show 1.96 Monte Carlo standard errors across 1,200 simulated datasets.</desc>
<defs><linearGradient id="bg" x1="0" y1="0" x2="1" y2="1"><stop stop-color="#07101f"/><stop offset="1" stop-color="#10213b"/></linearGradient></defs>
<rect width="1280" height="720" fill="url(#bg)"/>
<style>
text{font-family:ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,monospace;fill:#e5eefc}.eyebrow{font-size:15px;letter-spacing:2px;fill:#8fb7ff}.title{font-family:ui-sans-serif,system-ui,sans-serif;font-size:31px;font-weight:700}.subtitle{font-size:15px;fill:#a9bad3}.panel-title{font-size:17px;font-weight:700}.axis{font-size:12px;fill:#9fb0c9}.label{font-size:11px;fill:#c8d5e8}.legend{font-size:13px;fill:#c8d5e8}.note{font-size:13px;fill:#9fb0c9}
</style>
<text x="56" y="45" class="eyebrow">MONTE CARLO RESULT • 1,200 REPEATS PER CELL</text>
<text x="56" y="82" class="title">The correct resampling unit depends on the dependency structure</text>
<text x="56" y="108" class="subtitle">Points are empirical 95% interval coverage; vertical bars are ±1.96 Monte Carlo SE; horizontal reference is nominal 95%.</text>
${legend}
${panel(heterogeneity, 42, 158, 582, 480, "A  Fixed 40 repositories; heterogeneity changes")}
${panel(clusters, 656, 158, 582, 480, "B  Strong heterogeneity; repository count changes")}
<text x="56" y="678" class="note">Negative result: with only 10 repositories, the percentile repository bootstrap covered 90.4%—worse than task-level methods.</text>
<text x="56" y="700" class="note">Scope: hierarchical Bernoulli simulation, 8 tasks/repository, true mean effect +5 points; not an estimate of any named benchmark.</text>
</svg>\n`;

fs.writeFileSync(path.join(root, "clustered-agent-benchmark-uncertainty.svg"), svg);
console.log("wrote clustered-agent-benchmark-uncertainty.svg from results.json");

File: operator/diy-project-blogs/projects/prompt-cache-metric-audit/render-figure.py
import html
import json
import pathlib


PROJECT = pathlib.Path(__file__).resolve().parent
RESULTS = json.loads((PROJECT / "results.json").read_text())


def find_cell(design, alignment, capacity, ttl, cadence):
    return next(
        cell
        for cell in RESULTS["cells"]
        if cell["design"] == design
        and cell["alignment"] == alignment
        and cell["capacityFraction"] == capacity
        and cell["ttlMinutes"] == ttl
        and cell["meanInterarrivalSeconds"] == cadence
    )


alignments = ["large-popular", "independent", "small-popular"]
labels = {
    "large-popular": "Large prefixes popular",
    "independent": "Size and popularity independent",
    "small-popular": "Small prefixes popular",
}
colors = {
    "request": "#f6c453",
    "token": "#5eead4",
    "savings": "#8fb7ff",
}


def panel(x, y, width, height, capacity, ttl, cadence):
    left = x + 70
    right = x + width - 26
    top = y + 64
    bottom = y + height - 70
    chart_height = bottom - top
    group_width = (right - left) / len(alignments)
    marks = [
        f'<rect x="{x}" y="{y}" width="{width}" height="{height}" rx="10" '
        'fill="#111b32" stroke="#2b3b5c"/>',
        f'<text x="{x + 24}" y="{y + 35}" class="panel-title">'
        f'{int(capacity * 100)}% capacity • {ttl}m TTL • mean arrival {cadence}s</text>',
    ]
    for tick in [0.0, 0.2, 0.4, 0.6, 0.8]:
        tick_y = bottom - tick * chart_height
        marks.append(
            f'<line x1="{left}" y1="{tick_y}" x2="{right}" y2="{tick_y}" '
            'stroke="#344464" stroke-dasharray="4 5"/>'
        )
        marks.append(
            f'<text x="{left - 10}" y="{tick_y + 4}" class="axis" '
            f'text-anchor="end">{int(tick * 100)}%</text>'
        )

    for index, alignment in enumerate(alignments):
        cell = find_cell("observed-size", alignment, capacity, ttl, cadence)
        center = left + group_width * index + group_width / 2
        values = [
            ("request", cell["metrics"]["requestHitRate"]["mean"]),
            ("token", cell["metrics"]["tokenHitRate"]["mean"]),
            ("savings", cell["metrics"]["savingsRate"]["mean"]),
        ]
        for offset, (key, value) in enumerate(values):
            bar_width = 24
            bar_x = center + (offset - 1) * 31 - bar_width / 2
            bar_y = bottom - value * chart_height
            marks.append(
                f'<rect x="{bar_x}" y="{bar_y}" width="{bar_width}" '
                f'height="{bottom - bar_y}" rx="4" fill="{colors[key]}"/>'
            )
            marks.append(
                f'<text x="{bar_x + bar_width / 2}" y="{bar_y - 7}" '
                f'class="value" text-anchor="middle">{value * 100:.1f}</text>'
            )
        words = labels[alignment].split()
        first = " ".join(words[:2])
        second = " ".join(words[2:])
        marks.append(
            f'<text x="{center}" y="{bottom + 24}" class="label" '
            f'text-anchor="middle">{html.escape(first)}</text>'
        )
        if second:
            marks.append(
                f'<text x="{center}" y="{bottom + 41}" class="label" '
                f'text-anchor="middle">{html.escape(second)}</text>'
            )
    return "\n".join(marks)


legend = []
for index, (key, label) in enumerate(
    [
        ("request", "Request hit rate"),
        ("token", "Token-weighted hit rate"),
        ("savings", "Input-cost savings"),
    ]
):
    x = 165 + index * 280
    legend.append(
        f'<rect x="{x}" y="121" width="14" height="14" rx="3" '
        f'fill="{colors[key]}"/><text x="{x + 22}" y="133" '
        f'class="legend">{label}</text>'
    )

svg = f"""<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1280 720"
 width="1280" height="720" data-visual-quality="publication"
 data-text-fit="bounded" role="img" aria-labelledby="title desc">
<title id="title">Prompt-cache request hits, token hits, and cost savings diverge</title>
<desc id="desc">Two grouped bar panels compare request hit rate, token-weighted hit rate,
and modeled input-cost savings for workloads where large prefixes are popular,
prefix size and popularity are independent, or small prefixes are popular.
Values average 40 independently seeded traces of 12,000 requests.</desc>
<defs><linearGradient id="bg" x1="0" y1="0" x2="1" y2="1">
<stop stop-color="#07101f"/><stop offset="1" stop-color="#10213b"/>
</linearGradient></defs>
<rect width="1280" height="720" fill="url(#bg)"/>
<style>
text{{font-family:ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,monospace;fill:#e5eefc}}
.eyebrow{{font-size:15px;letter-spacing:2px;fill:#8fb7ff}}
.title{{font-family:ui-sans-serif,system-ui,sans-serif;font-size:31px;font-weight:700}}
.subtitle{{font-size:14px;fill:#a9bad3}}
.panel-title{{font-size:17px;font-weight:700}}
.axis{{font-size:12px;fill:#9fb0c9}}
.label{{font-size:11px;fill:#c8d5e8}}
.legend{{font-size:13px;fill:#c8d5e8}}
.value{{font-size:11px;font-weight:700}}
.note{{font-size:13px;fill:#9fb0c9}}
</style>
<text x="56" y="44" class="eyebrow">CONTROLLED CACHE SIMULATION • 40 TRACES PER CELL</text>
<text x="56" y="82" class="title">A high cache-hit rate can still save the wrong tokens</text>
<text x="56" y="106" class="subtitle">o200k token sizes from 52 committed articles; Zipf request popularity; whole-prefix LRU.</text>
{''.join(legend)}
{panel(42, 154, 582, 486, 0.2, 5, 3)}
{panel(656, 154, 582, 486, 0.2, 30, 180)}
<text x="56" y="678" class="note">Modeled billing: cache write 1.25× uncached input; cache read 0.10×; hits refresh TTL.</text>
<text x="56" y="700" class="note">Scope: metric and policy behavior under the declared trace model—not provider latency or production cache residency.</text>
</svg>
"""
(PROJECT / "prompt-cache-metric-audit.svg").write_text(svg)
print("wrote prompt-cache-metric-audit.svg from results.json")
